crates/wasmtime/src/component/func/typed.rs

```diff
@@ -0,0 +1,1924 @@
+// * The bounds-checks in `Memory::get` were removed (replaced with unsafe
+//   indexing)
+//
+// Even then though this didn't correctly vectorized for `Vec<u8>`. It's not
+// entirely clear why but it appeared that it's related to reloading the base
+// pointer fo memory (I guess from `Memory` itself?). Overall I'm not really
+// clear on what's happening there, but this is surely going to be a performance
+// bottleneck in the future.
+fn lower_list<T, U>(mem: &mut Memory<'_, U>, list: &[T]) -> Result<(usize, usize)>
+where
+    T: ComponentValue,
+{
+    let elem_size = T::size();
+    let size = list
+        .len()
+        .checked_mul(elem_size)
+        .ok_or_else(|| anyhow::anyhow!("size overflow copying a list"))?;
+    let mut ptr = mem.realloc(0, 0, T::align(), size)?;
+    for item in list {
+        item.store(mem, ptr)?;
+        ptr += elem_size;
+    }
+    Ok((ptr, list.len()))
+}
+
+impl<'a, T: ComponentValue> Cursor<'a, Vec<T>> {
+    /// Returns the item length of this vector
+    pub fn len(&self) -> usize {
+        // FIXME: needs memory64 treatment
+        u32::from_le_bytes(self.item_bytes()[4..].try_into().unwrap()) as usize
+    }
+
+    /// Returns an iterator over the elements of this vector.
+    ///
+    /// The returned iterator is an exact-size iterator and is of length
+    /// `self.len()`. Note that the iterator is also an iterator of [`Cursor`]
+    /// types representing that the desired values all continue to live in wasm
+    /// linear memory.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if this list's pointer/length combination is
+    /// out-of-bounds, or if the length times the element size is too large to
+    /// fit in linear memory.
+    pub fn iter(&self) -> Result<impl ExactSizeIterator<Item = Cursor<'a, T>> + '_> {
+        let (ptr, len) = {
+            let ptr_and_len = self.item_bytes();
+            // FIXME: needs memory64 treatment
+            let ptr = u32::from_le_bytes(ptr_and_len[..4].try_into().unwrap());
+            let len = u32::from_le_bytes(ptr_and_len[4..].try_into().unwrap());
+            (usize::try_from(ptr)?, usize::try_from(len)?)
+        };
+        len.checked_mul(T::size())
+            .and_then(|byte_len| self.all_memory().get(ptr..)?.get(..byte_len))
+            .ok_or_else(|| anyhow::anyhow!("list out of bounds"))?;
+
+        Ok((0..len).map(move |i| {
+            // The `move_to` function is not safe because `Cursor` is a static
+            // proof that the offset/length is in-bounds. This bounds-check,
+            // however, was just performed above so we know that the offset is
+            // indeed valid, meaning this `unsafe` should be ok.
+            unsafe { self.move_to(ptr + T::size() * i) }
+        }))
+    }
+}
+
+#[inline]
+const fn align_to(a: usize, align: u32) -> usize {
+    let align = align as usize;
+    (a + (align - 1)) & !(align - 1)
+}
```

Comment:
Debug assert that `align` is a power of two (which the `&` stuff relies on, I believe)?

Alternatively, just use division (which will work with non-powers-of-two) and let LLVM clean up the generated code.